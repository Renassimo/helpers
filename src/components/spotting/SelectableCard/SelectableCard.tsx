import React, { ReactNode } from 'react';

import Card from '@mui/material/Card';
import Box from '@mui/material/Box';
import IconButton from '@mui/material/IconButton';

import RadioButtonUncheckedIcon from '@mui/icons-material/RadioButtonUnchecked';
import CheckCircleIcon from '@mui/icons-material/CheckCircle';

const SelectedIcon = () => <CheckCircleIcon color="primary" />;

const UnselectedIcon = ({
  isAnySelected = false,
}: {
  isAnySelected?: boolean;
}) => (isAnySelected ? <RadioButtonUncheckedIcon /> : <CheckCircleIcon />);

const SelectIconButton = ({
  isAnySelected = false,
  selected = false,
  onClick,
}: {
  isAnySelected?: boolean;
  selected?: boolean;
  onClick: () => void;
}) => (
  <IconButton onClick={onClick}>
    {selected ? (
      <SelectedIcon />
    ) : (
      <UnselectedIcon isAnySelected={isAnySelected} />
    )}
  </IconButton>
);

const CardChecker = ({
  isAnySelected = false,
  selected = false,
  toggleSelect,
}: {
  isAnySelected?: boolean;
  selected?: boolean;
  toggleSelect: () => void;
}) => {
  const isInvisible = !isAnySelected && !selected;
  return (
    <Box
      sx={{
        position: 'absolute',
        right: '0',
        opacity: isInvisible ? '0' : '1',
        transition: 'opacity 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
      }}
      className="group-check-button-box"
    >
      <SelectIconButton
        isAnySelected={isAnySelected}
        selected={selected}
        onClick={toggleSelect}
      />
    </Box>
  );
};

const SelectableCard = ({
  children,
  selectable = false,
  selected = false,
  isAnySelected = false,
  toggleSelect,
}: {
  children: ReactNode;
  selectable?: boolean;
  selected?: boolean;
  isAnySelected?: boolean;
  toggleSelect: () => void;
}) => {
  const isSelected = selectable && selected;
  return (
    <Card
      sx={{
        position: 'relative',
        boxShadow: isSelected ? '0px 0px 2px 2px #1976d2' : 'none',
        '&:hover': {
          '& .group-check-button-box': { opacity: '1' },
        },
      }}
    >
      {selectable && (
        <CardChecker
          isAnySelected={isAnySelected}
          selected={isSelected}
          toggleSelect={toggleSelect}
        />
      )}
      {children}
    </Card>
  );
};

export default SelectableCard;
