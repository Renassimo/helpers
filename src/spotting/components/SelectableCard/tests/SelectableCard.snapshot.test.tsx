import renderWithTheme from '@/tests/helpers';

import SelectableCard from '@/spotting/components/SelectableCard';

describe('SelectableCard snapshot', () => {
  const mockedToggleSelect = jest.fn();

  test('renders successfully', () => {
    // Arrange
    // Act
    const { container } = renderWithTheme(
      <SelectableCard toggleSelect={mockedToggleSelect}>
        SelectableCard
      </SelectableCard>
    );
    // Assert
    expect(container).toMatchSnapshot();
  });

  describe('when selectable', () => {
    describe('and selected', () => {
      describe('and there are other selected', () => {
        test('renders successfully', () => {
          // Arrange
          // Act
          const { container } = renderWithTheme(
            <SelectableCard
              selectable
              selected
              isAnySelected
              toggleSelect={mockedToggleSelect}
            >
              SelectableCard
            </SelectableCard>
          );
          // Assert
          expect(container).toMatchSnapshot();
        });
      });

      describe('and there no are other selected', () => {
        test('renders successfully', () => {
          // Arrange
          // Act
          const { container } = renderWithTheme(
            <SelectableCard
              selectable
              selected
              isAnySelected={false}
              toggleSelect={mockedToggleSelect}
            >
              SelectableCard
            </SelectableCard>
          );
          // Assert
          expect(container).toMatchSnapshot();
        });
      });
    });

    describe('and not selected', () => {
      describe('and there are other selected', () => {
        test('renders successfully', () => {
          // Arrange
          // Act
          const { container } = renderWithTheme(
            <SelectableCard
              selectable
              selected={false}
              isAnySelected
              toggleSelect={mockedToggleSelect}
            >
              SelectableCard
            </SelectableCard>
          );
          // Assert
          expect(container).toMatchSnapshot();
        });
      });

      describe('and there no are other selected', () => {
        test('renders successfully', () => {
          // Arrange
          // Act
          const { container } = renderWithTheme(
            <SelectableCard
              selectable
              selected={false}
              isAnySelected={false}
              toggleSelect={mockedToggleSelect}
            >
              SelectableCard
            </SelectableCard>
          );
          // Assert
          expect(container).toMatchSnapshot();
        });
      });
    });
  });
});
