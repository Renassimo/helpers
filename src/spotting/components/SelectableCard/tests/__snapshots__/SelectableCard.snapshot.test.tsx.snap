// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectableCard snapshot renders successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-171dp3o-MuiPaper-root-MuiCard-root"
  >
    SelectableCard
  </div>
</div>
`;

exports[`SelectableCard snapshot when selectable and not selected and there are other selected renders successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-171dp3o-MuiPaper-root-MuiCard-root"
  >
    <div
      class="group-check-button-box MuiBox-root css-36ngln"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="RadioButtonUncheckedIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm0 18c-4.42 0-8-3.58-8-8s3.58-8 8-8 8 3.58 8 8-3.58 8-8 8z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    SelectableCard
  </div>
</div>
`;

exports[`SelectableCard snapshot when selectable and not selected and there no are other selected renders successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-171dp3o-MuiPaper-root-MuiCard-root"
  >
    <div
      class="group-check-button-box MuiBox-root css-iy8yss"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    SelectableCard
  </div>
</div>
`;

exports[`SelectableCard snapshot when selectable and selected and there are other selected renders successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-s69h3n-MuiPaper-root-MuiCard-root"
  >
    <div
      class="group-check-button-box MuiBox-root css-36ngln"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
          data-testid="CheckCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    SelectableCard
  </div>
</div>
`;

exports[`SelectableCard snapshot when selectable and selected and there no are other selected renders successfully 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-s69h3n-MuiPaper-root-MuiCard-root"
  >
    <div
      class="group-check-button-box MuiBox-root css-36ngln"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-colorPrimary MuiSvgIcon-fontSizeMedium css-ci5apu-MuiSvgIcon-root"
          data-testid="CheckCircleIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm-2 15-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
    SelectableCard
  </div>
</div>
`;
